### PR TITLE
Support for cancelling queries

### PR DESCRIPTION
**Pull request by macobo**
See #77

### CODE REVIEW

**Comment by macobo** [comment]
Note that the test failures are the same as in the current master!

File: pgcli/pgexecute.py [hunk 1]
```diff
@@ -11,6 +12,9 @@
 psycopg2.extensions.register_type(psycopg2.extensions.UNICODE)
 psycopg2.extensions.register_type(psycopg2.extensions.UNICODEARRAY)
 
+# When running a query, make pressing CTRL+C raise a KeyboardInterrupt
+psycopg2.extensions.set_wait_callback(psycopg2.extras.wait_select)
```
Comment (amjith) [inline pgcli/pgexecute.py]: This is interesting. It works as expected but I don't understand how. Reading the documentation says this waits for data to be available, but during the wait time does it automatically capture Ctrl+C and convert it to KeyboardInterrupt? Is this a behavior of `select()` or `wait_select()`?

File: pgcli/main.py [hunk 2]
```diff
@@ -156,6 +156,10 @@ def cli(database, user, host, port, prompt_passwd, never_prompt):
                     _logger.debug("status: %r", status)
                     output.extend(format_output(rows, headers, status))
                 click.echo_via_pager('\n'.join(output))
+            except KeyboardInterrupt:
+                pgexecute.reconnect()
```
Comment (amjith) [inline pgcli/main.py]: Would it be prudent to call [psycopg2.connection.cancel()](http://initd.org/psycopg/docs/connection.html#connection.cancel)? Just to make sure we're not leaving a long running query on the server and just abandoning a connection.

Comment (macobo) [inline pgcli/main.py]: Good catch!

My first stab at this showed that the connection was automatically cancelled by `psycopg2.extras.wait_select`.

Looking at the implementation of `wait_select` though, I am not 100% sure WHY it closes the connection. Will test and give you an update.

File: pgcli/pgexecute.py (hunk 1 above)
Comment (macobo) [inline pgcli/pgexecute.py]: This was explained here: http://initd.org/psycopg/articles/2014/07/20/cancelling-postgresql-statements-python/

Note that wait_select is supposed to carry out the same role as the function defined in that article.

Comment (amjith) [inline pgcli/pgexecute.py]: Yeah I noticed that in the article after I left the comment. Now I'm trying to find the source for wait_select() to check the difference between wait_select() and wait_select_inter().

Comment (macobo) [inline pgcli/pgexecute.py]: Updated the PR with link to the article.

The function can be found [here](https://github.com/psycopg/psycopg2/blob/df9fbc515d6f586e84d7853965f2bc544177ac6a/lib/extras.py#L566). It is the same function as the one from the article, minus handling KeyboardInterrupt.

File: pgcli/main.py (hunk 2 above)
Comment (macobo) [inline pgcli/main.py]: Updated to explicitly close the connection.

**Comment by amjith** [comment]
Thanks for doing this. 

:beetle:

**Pull request merged**